src/libpmempool/sync.c

```diff
@@ -90,6 +90,194 @@ validate_args(struct pool_set *set)
+				rep->remote->node_addr);
+			return -1;
+		}
+	} else if (rep_h->remote) {
+		LOG(10,
+			"Reading data (offset %zu length %zu) from remote node -- '%s' on '%s'",
+			off, len,
+			rep_h->remote->pool_desc,
+			rep_h->remote->node_addr);
+
+		ret = Rpmem_read(rep_h->remote->rpp, dst_addr, off, len, 0);
+		if (ret) {
+			LOG(1,
+				"Reading data from remote node failed -- '%s' on '%s'",
+				rep_h->remote->pool_desc,
+				rep_h->remote->node_addr);
+			return -1;
+		}
+	} else {
+		LOG(10,
+			"Copying data (offset %zu length %zu) from local replica -- '%s'",
+			off, len, rep_h->part[0].path);
+
+		/* copy all data */
+		memcpy(dst_addr, src_addr, len);
+		util_persist(part->is_dev_dax, dst_addr, len);
+	}
+
+	return 0;
+}
+
+/*
+ * sync_recreate_header -- (internal) recreate the header
+ */
+static int
+sync_recreate_header(struct pool_set *set, unsigned r, unsigned p,
+			struct pool_hdr *src_hdr)
+{
+	struct pool_attr attr;
+	util_pool_hdr2attr(&attr, src_hdr);
+	if (util_header_create(set, r, p, &attr, 1) != 0) {
+		LOG(1, "part headers create failed for replica %u part %u",
+			r, p);
+		errno = EINVAL;
+		return -1;
+	}
+
+	return 0;
+}
+
+/*
+ * sync_badblocks_data -- (internal) clear bad blocks in replica
+ */
+static int
+sync_badblocks_data(struct pool_set *set, unsigned healthy_replica,
+			struct poolset_health_status *set_hs)
+{
+	LOG(3, "set %p, set_hs %p", set, set_hs);
+
+	struct pool_replica *rep_h = REP(set, healthy_replica);
+	struct pool_hdr *src_hdr = HDR(REP(set, healthy_replica), 0);
+
+	size_t hdrsize = (set->options & (OPTION_SINGLEHDR | OPTION_NOHDRS)) ?
+				0 : POOL_HDR_SIZE;
+
+	for (unsigned r = 0; r < set->nreplicas; ++r) {
+		struct pool_replica *rep = REP(set, r);
+		struct replica_health_status *rep_hs = set_hs->replica[r];
+
+		if (!(rep_hs->flags & HAS_BAD_BLOCKS)) {
+			/* skip replicas with no bad blocks */
+			continue;
+		}
+
+		LOG(10, "Replica %u HAS bad blocks", r);
+
+		for (unsigned p = 0; p < rep->nparts; ++p) {
+
+			struct part_health_status *phs = &rep_hs->part[p];
+
+			if (!(phs->flags & HAS_BAD_BLOCKS)) {
+				/* skip parts with no bad blocks */
+				continue;
+			}
+
+			if (phs->bbs.bb_cnt == 0 || phs->bbs.bbv == NULL) {
+				/* mark part as having no bad blocks */
+				replica_mark_part_no_badblocks(r, p, set_hs);
+				continue;
+			}
+
+			LOG(10, "Replica %u part %u HAS %u bad blocks",
+				r, p, phs->bbs.bb_cnt);
+
+			const struct pool_set_part *part = &rep->part[p];
+			size_t part_off = replica_get_part_offset(set, r, p);
+
+			for (unsigned i = 0; i < phs->bbs.bb_cnt; i++) {
+				LOG(10,
+					"fixing bad block #%i: offset %llu, length %u",
+					i,
+					phs->bbs.bbv[i].offset,
+					phs->bbs.bbv[i].length);
+
+				size_t off = phs->bbs.bbv[i].offset;
+				size_t len = phs->bbs.bbv[i].length;
+
+				if (off < POOL_HDR_SIZE &&
+				    sync_recreate_header(set, r, p, src_hdr))
+					return -1;
+
+				/* parts #>0 are mapped without the header */
+				if (p > 0) {
```

Comment:
I feel like we are doing this calculation in too many places already. Is there a way in which we could move all of this stuff into a couple of functions that would be used everywhere?
We already had a bug where on of the places where we removed the header size had a bug and we had a problem to locate that place...